clientLibrary/src/main/scala/za/co/absa/loginclient/authorization/JWTDecoderProvider.scala

```diff
@@ -0,0 +1,110 @@
+import za.co.absa.loginclient.tokenRetrieval.service.RetrieveToken
+
+import java.text.SimpleDateFormat
+import scala.concurrent.duration.FiniteDuration
+
+/**
+ * This class is used to decode JWT tokens.
+ * It can be used to create decoders from publicKeys which decode access tokens.
+ * It can also be used to verify access tokens.
+ * Optionally, you may set it to periodically update the decoder with a new publickey.
+ * @param publicKeyEndpoint The endpoint to retrieve the public key from.
+ * @param refreshPeriod The period at which to refresh the public key. Optional Parameter.
+ */
+
+case class JWTDecoderProvider(publicKeyEndpoint : String, refreshPeriod : Option[FiniteDuration] = None) extends JwtDecoder {
+
+  private val logger = LoggerFactory.getLogger(this.getClass)
+  private val tokenRetrieval = RetrieveToken(publicKeyEndpoint)
+  @volatile private var decoder: JwtDecoder = createDecoder(tokenRetrieval.getPublicKey)
+  if (refreshPeriod.nonEmpty) scheduleKeyRefresh()
+
+  private def scheduleKeyRefresh(): Unit = {
+    val scheduler = Executors.newSingleThreadScheduledExecutor(r => {
+      val t = new Thread(r)
+      t.setDaemon(true)
+      t
+    })
+    scheduler.scheduleAtFixedRate(() => {
+      refreshDecoder()
+    }, refreshPeriod.get.toMillis,
+      refreshPeriod.get.toMillis,
+      TimeUnit.MILLISECONDS
+    )
+  }
+
+  private def refreshDecoder(): Unit = {
+    try {
+      logger.info(s"Refreshing Public Key from $publicKeyEndpoint")
+      decoder = createDecoder(tokenRetrieval.getPublicKey)
+      logger.info(s"Successfully refreshed Public Key from $publicKeyEndpoint" +
+        s"next refresh will occur in ${refreshPeriod.get.toSeconds} seconds")
+    }
+    catch {
+      case e: Throwable =>
+        logger.error(s"Error occurred retrieving and decoding Public Key from $publicKeyEndpoint " +
+          s"will attempt to retrieve Public Key again in ${refreshPeriod.get.toSeconds} seconds", e)
+    }
+  }
+
+  private def createDecoder(publicKey: PublicKey): JwtDecoder = {
+    val publicKeyBytes = Base64.getDecoder.decode(publicKey.token)
+    val publicKeySpec = new X509EncodedKeySpec(publicKeyBytes)
+    val encodedPublicKey = KeyFactory.getInstance("RSA").generatePublic(publicKeySpec).asInstanceOf[RSAPublicKey]
+    NimbusJwtDecoder.withPublicKey(encodedPublicKey).build()
+  }
+
+  override def decode(token: String): Jwt = decoder.decode(token)
+
+  def decode(token: Token): Jwt = decode(token.token)
+
+  def verifyAccessToken(token: AccessToken): Map[String, Any] = {
+    try {
+      logger.info("Verifying access token")
+      val jwt = decode(token)
+      val claims = getAllClaims(jwt)
+      val dateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss")
+      if(dateFormat.parse(claims("exp").toString).getTime < System.currentTimeMillis())
+        throw new Exception("Token has expired")
+      if(claims("type").toString != "access")
```

Comment:
@dk1844 What do you think? So the main thing our own code offer is the ability to poll and like Kevin mentioned checking the Access Token, so do you think it would be better for people to just use Nimbus directly or to use a wrapper like we have now that's specific to Login Service?

Reply:
I'm siding with @kevinwallimann's suggestion - we already import Nimbus dependencies anyway. The only concern I have there if the simplicity and readability will not suffer. Otherwise I find relying on a stable library to be a good idea, generally.

Reply:
Ok, so the Decoder Code has been removed and Nimbus is being used directly in the example code. Verifying claims has been moved to the claims parser, so please let me know if this is better.

Reply:
I don't think that the code we ended up with now is what @kevinwallimann meant. I have looked at  [NimbusJwtDecoder.JwkSetUriJwtDecoderBuilder](https://github.com/spring-projects/spring-security/blob/main/oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/NimbusJwtDecoder.java#L262) and it directly takes the argument of jwks path, so for us, it would be `s"$host/token/public-key-jwks"`. 

I don't think that it should all be done in the `TokenRetrievalService` (or newly named `TokenRetrievalClient`) - this client should include the NimbusJwtDecoder with the JWKs path already, so it should be able check the JWT against the public key internally, besides the current functionality.

As testing of `TokenRetrievalClient` is concerned, we may provide testing decoder based on a known public key via `NimbusJwtDecoder.withPublicKey(encodedPublicKey).build()` instead of `NimbusJwtDecoder.withJwkSetUri("xxx").build()` that would get used normally (or we could allow both initialization variant for this class?)